### PR TITLE
Update dependencies and Rust that's used

### PR DESCRIPTION
**Pull request by alexcrichton**
* Picks up a few breaking changes to dependency APIs, notably the tar-rs changes
  and git2-rs changes.
* Builds should now work with VS 2015

### CODE REVIEW

File: src/cargo/sources/git/utils.rs [hunk 1]
```diff
@@ -368,8 +368,13 @@ fn with_authentication<T, F>(url: &str, cfg: &git2::Config, mut f: F)
     let mut cred_helper = git2::CredentialHelper::new(url);
     cred_helper.config(cfg);
     let mut cred_error = false;
+    let mut called = 0;
     let ret = f(&mut |url, username, allowed| {
-        let creds = if allowed.contains(git2::SSH_KEY) ||
+        called += 1;
```
Comment (huonw) [inline src/cargo/sources/git/utils.rs]: This is kinda strange, could you add a comment?

Comment (alexcrichton) [inline src/cargo/sources/git/utils.rs]: Good idea, comment added!

**Pull request merged**